v-next/hardhat/test/internal/cli/main.ts

```diff
@@ -57,7 +62,199 @@ async function getTasksAndHreEnvironment(
   };
 }
 
+async function getTasksAndSubtaskResults(
+  configFileName: string = "hardhat.config.ts",
+): Promise<boolean[]> {
+  // To ensure that one or more tasks have been executed, each task will modify an array of boolean values, initially set to false.
+  // This function imports that array, allowing the tests to verify if the tasks have been executed.
+  // If a boolean flag is true, it indicates that the corresponding task (or a specific part of it) has been executed.
+  // The array is set in the hardhat.config.ts file of the fixture project.
+  return (
+    await import(pathToFileURL(`${process.cwd()}/${configFileName}`).toString())
+  ).results;
+}
+
 describe("main", function () {
+  describe("main", function () {
+    afterEach(function () {
+      resetHardhatRuntimeEnvironmentSingleton();
+      mock.reset();
+    });
+
+    describe("version", function () {
+      useFixtureProject("cli/parsing/base-project");
+
+      // TODO: as soon as the 'version task' is done, this test should be updated
+      it.todo(
+        "should print the version and instantly return",
+        async function () {
+          const m = mock.method(console, "log", () => {});
+          const spy = mock.fn(importUserConfig);
```

Comment:
I am not sure about this level of mocking. I don't think we should do it at the level of `console.log`.

I think I would prefer an approach where `main` takes additional optional parameters for logging and for importing a user config.

In tests we inject functions or mock spys. In `cli.ts` we pass in the actual implementation.

What do we think?

Reply:
@alcuadrado what do you think? I'm asking about this specific scenario, but also in general. How should we approach spying and mocking in the new v3 code?

Reply:
I think I prefer @kanej's approach, as this level of mocking is assuming a certain implementation.

I think `main` can take an optional `print = console.sol` param, or something along those lines.

The change here would be small.

Reply:
Done in this commit with readability improvement for tests too (like mentioned in your last commit @alcuadrado): https://github.com/NomicFoundation/hardhat/pull/5380/commits/213f009b955f89a9f5e3565d5c0fb02b266196e3